feat(Order): A monotone function `f` maps `sSup s` to `f (sSup s)`

dupuisf opened the pull request:
This PR shows several lemmas related to the PR title.

Note that we already have several lemmas in this direction, but they have very strong hypotheses (like `Monotone f` when `MonotoneOn f s` would do). 

Also, there's quite a bit of code duplication: most notably, `Monotone.map_sSup_of_continuousAt'` is the same as `Monotone.map_csSup_of_continuousAt` except that there are some autoparams in the former (and likewise for the rest of the collection). I've kept everything, but I could definitely be convinced to delete one version.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

dupuisf commented:
> _Any_ function `f` maps `sSup s` to `f (sSup s)` 😁 Mind fixing the title?

I added "continuous" -- I think if people are interested in the details they can have a look at the code :-)

YaelDillies commented:
That's not really what I meant. Let me fix it

YaelDillies commented on Mathlib/Topology/Order/Monotone.lean:
```diff
@@ -28,12 +28,20 @@ variable [ConditionallyCompleteLinearOrder α] [TopologicalSpace α] [OrderTopol
 the supremum of the image of this set. -/
-theorem Monotone.map_sSup_of_continuousAt' {f : α → β} {A : Set α} (Cf : ContinuousAt f (sSup A))
-    (Mf : Monotone f) (A_nonemp : A.Nonempty) (A_bdd : BddAbove A := by bddDefault) :
+theorem MonotoneOn.map_sSup_of_continuousWithinAt' {f : α → β} {A : Set α}
+    (Cf : ContinuousWithinAt f A (sSup A))
+    (Mf : MonotoneOn f A) (A_nonemp : A.Nonempty) (A_bdd : BddAbove A := by bddDefault) :
     f (sSup A) = sSup (f '' A) :=
```
For lemmas like this, stating the property of some set depending on some function property (continuity, monotonicity, convexity, differentiability...), I tend to suggest stating the required function property assumptions on a *superset* of the set in the conclusion. This is useful, because usually the set `A` you're trying to plug in is *not* the domain on which the function respect the property, which would rather be an open half-line, or the complement of a singleton or what not.

Here, concretely, I would do
```suggestion
theorem MonotoneOn.map_sSup_of_continuousWithinAt' {f : α → β} {A B : Set α} (hAB : A ⊆ B)
    (Cf : ContinuousWithinAt f B (sSup A))
    (Mf : MonotoneOn f B) (A_nonemp : A.Nonempty) (A_bdd : BddAbove A := by bddDefault) :
    f (sSup A) = sSup (f '' A) :=
```
or even
```suggestion
theorem MonotoneOn.map_sSup_of_continuousWithinAt' {f : α → β} {A B C : Set α} (hAB : A ⊆ B)
    (hAC : A ⊆ C) (Cf : ContinuousWithinAt f B (sSup A))
    (Mf : MonotoneOn f C) (A_nonemp : A.Nonempty) (A_bdd : BddAbove A := by bddDefault) :
    f (sSup A) = sSup (f '' A) :=
```
but that's probably overengineered

dupuisf replied:
Fair enough, I went with the first option.

YaelDillies replied:
Actually Bhavik disagrees that this is worth doing except for things like `ConvexOn` (because `ConvexOn.subset` is so annoying to use), and I am somewhat convinced by this argument. Would you mind rolling back?

dupuisf replied:
No problem -- I confess I was a bit skeptical as well, when constructing a proof using `refine` I think it's a bit easier to deal with the subset thing afterwards.

YaelDillies reviewed (commented):
Indeed I don't think it's worth duplicating lemmas when the sole difference is auto-params. If auto-params are causing issues, then we should improve the user experience rather than duplicating lemmas

dupuisf commented:
> Indeed I don't think it's worth duplicating lemmas when the sole difference is auto-params. If auto-params are causing issues, then we should improve the user experience rather than duplicating lemmas

I removed the duplicates. I kept the version with the autoparam and added deprecated aliases.

YaelDillies reviewed (approved):
Thanks!

maintainer merge

jcommelin reviewed (commented):
Thanks :tada:

bors merge
